ddtrace/tracer/textmap.go

```diff
@@ -114,74 +115,188 @@ func NewPropagator(cfg *PropagatorConfig) Propagator {
+		}
+	}
+
+	// If all the styles were invalid/unknown, then revert to default behavior.
+	if len(injectors) == 0 {
+		return []textMapPropagator{dd}
+	}
+	return injectors
+}
+
+func makeExtractors(cfg *PropagatorConfig) []textMapPropagator {
+	dd := &datadogPropagator{cfg}
+	b3 := &b3Propagator{}
+
+	ps := os.Getenv("DD_PROPAGATION_STYLE_EXTRACT")
+	if ps == "" {
+		return []textMapPropagator{dd}
+	}
+	styles := propagationStyleList(ps)
+
+	var extractors []textMapPropagator
+	for _, v := range styles {
+		switch v {
+		case "Datadog":
+			extractors = append(extractors, dd)
+		case "B3":
+			extractors = append(extractors, b3)
+		default:
+			// TODO: consider logging something for invalid/unknown styles.
+		}
+	}
+
+	// If all the styles were invalid/unknown, then revert to default behavior.
+	if len(extractors) == 0 {
+		return []textMapPropagator{dd}
+	}
+	return extractors
+}
+
+// propagationStyleList splits a string containing propagation styles
+// separated by space or comma, and returns a []string containing
+// only the propagation styles.
+func propagationStyleList(s string) []string {
+	f := func(r rune) bool {
+		switch r {
+		case ' ', ',':
+			return true
+		}
+		return false
+	}
+	return strings.FieldsFunc(s, f)
+}
 
 // Inject defines the Propagator to propagate SpanContext data
 // out of the current process. The implementation propagates the
 // TraceID and the current active SpanID, as well as the Span baggage.
 func (p *propagator) Inject(spanCtx ddtrace.SpanContext, carrier interface{}) error {
-	switch v := carrier.(type) {
+	switch c := carrier.(type) {
 	case TextMapWriter:
-		return p.injectTextMap(spanCtx, v)
+		// Apply all of the injectors.
+		for _, v := range p.injectors {
+			err := v.injectTextMap(spanCtx, c)
+			if err != nil {
+				return err
+			}
+		}
+		return nil
 	default:
 		return ErrInvalidCarrier
 	}
 }
 
-func (p *propagator) injectTextMap(spanCtx ddtrace.SpanContext, writer TextMapWriter) error {
+// Extract implements Propagator.
+func (p *propagator) Extract(carrier interface{}) (ddtrace.SpanContext, error) {
+	switch c := carrier.(type) {
+	case TextMapReader:
+		// Try each extractor. The first to successfully produce a context
+		// will get returned.
+		for _, v := range p.extractors {
+			ctx, err := v.extractTextMap(c)
+			if ctx != nil {
+				return ctx, nil
+			}
+			// Special treatment for "context not found"
+			if err == ErrSpanContextNotFound {
+				continue
+			}
+			return nil, err
+		}
+		return nil, ErrSpanContextNotFound
+	default:
+		return nil, ErrInvalidCarrier
+	}
+}
+
+// textMapPropagator is used to inject and extract span contexts.
+type textMapPropagator interface {
+	injectTextMap(spanCtx ddtrace.SpanContext, writer TextMapWriter) error
+	extractTextMap(reader TextMapReader) (ddtrace.SpanContext, error)
+}
+
+// datadogPropagator implements textMapPropagator and injects/extracts span contexts
+// using datadog headers.
+type datadogPropagator struct {
```

Comment:
Generally in all our Go repos we've learned (the hard way) to do our best to avoid using the word "datadog" in identifiers. I'd like to encourage doing the same here and just calling this `textMapPropagator` (see next comments).